File: examples/python/diff_rpms.py
"""Compare the headers of two RPM packages and report differences."""

import argparse

from rpm_rs import PackageMetadata, SignatureTag, Tag


def tag_name(tag_id, enum_cls):
    try:
        return f"{enum_cls(tag_id).name} [{tag_id}]"
    except ValueError:
        return str(tag_id)


def format_value(val):
    if isinstance(val, str):
        return repr(val)
    if isinstance(val, bytes):
        if len(val) <= 16:
            return val.hex(" ")
        return f"[{len(val)} bytes]"
    if isinstance(val, list):
        if len(val) == 1:
            return repr(val[0])
        if len(val) <= 8:
            return repr(val)
        item_type = type(val[0]).__name__ if val else "?"
        return f"[{len(val)} {item_type}s]"
    return repr(val)


def diff_headers(label, a_entries, b_entries, enum_cls):
    all_tags = sorted(set(a_entries) | set(b_entries))

    lines = []
    count = 0
    for tag in all_tags:
        a_val = a_entries.get(tag)
        b_val = b_entries.get(tag)
        name = tag_name(tag, enum_cls)

        if a_val is not None and b_val is None:
            lines.append(f"  - {name} (only in first)")
            count += 1
        elif a_val is None and b_val is not None:
            lines.append(f"  + {name} (only in second)")
            count += 1
        elif a_val != b_val:
            lines.append(f"  ~ {name}:")
            lines.append(f"    < {format_value(a_val)}")
            lines.append(f"    > {format_value(b_val)}")
            count += 1

    if count == 0:
        print(f"\n{label}: identical")
    else:
        print(f"\n{label}:")
        for line in lines:
            print(line)
        print(f"\n  {count} difference(s)")


parser = argparse.ArgumentParser(description=__doc__)
parser.add_argument("rpm_a", help="First RPM file")
parser.add_argument("rpm_b", help="Second RPM file")
args = parser.parse_args()

a = PackageMetadata.open(args.rpm_a)
b = PackageMetadata.open(args.rpm_b)

print("Comparing:")
print(f"  A: {a.nevra()}")
print(f"  B: {b.nevra()}")

diff_headers(
    "Signature header",
    a.signature.get_all_entries(),
    b.signature.get_all_entries(),
    SignatureTag,
)

diff_headers(
    "Main header",
    a.header.get_all_entries(),
    b.header.get_all_entries(),
    Tag,
)

File: examples/diff-rpms.rs
use std::collections::BTreeMap;
use std::path::PathBuf;

use clap::Parser;
use rpm::{IndexData, PackageMetadata, display_tag_id};

/// Compare the headers of two RPM packages and report differences.
#[derive(Parser)]
#[command(version)]
struct Args {
    /// Path to the first RPM file
    a: PathBuf,

    /// Path to the second RPM file
    b: PathBuf,
}

fn format_value(data: &IndexData) -> String {
    match data {
        IndexData::Null => "Null".to_string(),
        IndexData::StringTag(s) => format!("{s:?}"),
        IndexData::StringArray(v) | IndexData::I18NString(v) => {
            if v.len() == 1 {
                format!("{:?}", v[0])
            } else {
                format!("[{} strings]", v.len())
            }
        }
        IndexData::Bin(v) | IndexData::Char(v) | IndexData::Int8(v) => {
            if v.len() <= 16 {
                format!("{v:02x?}")
            } else {
                format!("[{} bytes]", v.len())
            }
        }
        IndexData::Int16(v) => format!("{v:?}"),
        IndexData::Int32(v) => format!("{v:?}"),
        IndexData::Int64(v) => format!("{v:?}"),
    }
}

fn diff_headers(label: &str, a_entries: Vec<(u32, IndexData)>, b_entries: Vec<(u32, IndexData)>) {
    let a_map: BTreeMap<u32, IndexData> = a_entries.into_iter().collect();
    let b_map: BTreeMap<u32, IndexData> = b_entries.into_iter().collect();

    let mut all_tags: Vec<u32> = a_map.keys().chain(b_map.keys()).copied().collect();
    all_tags.sort();
    all_tags.dedup();

    let mut diffs = 0;

    for tag in &all_tags {
        match (a_map.get(tag), b_map.get(tag)) {
            (Some(_), None) => {
                if diffs == 0 {
                    println!("\n{label}:");
                }
                diffs += 1;
                println!("  - {} (only in first)", display_tag_id(*tag));
            }
            (None, Some(_)) => {
                if diffs == 0 {
                    println!("\n{label}:");
                }
                diffs += 1;
                println!("  + {} (only in second)", display_tag_id(*tag));
            }
            (Some(a_val), Some(b_val)) if a_val != b_val => {
                if diffs == 0 {
                    println!("\n{label}:");
                }
                diffs += 1;
                println!("  ~ {}:", display_tag_id(*tag));
                println!("    < {}", format_value(a_val));
                println!("    > {}", format_value(b_val));
            }
            _ => {}
        }
    }

    if diffs == 0 {
        println!("\n{label}: identical");
    } else {
        println!("\n  {diffs} difference(s)");
    }
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let args = Args::parse();

    let a = PackageMetadata::open(&args.a)?;
    let b = PackageMetadata::open(&args.b)?;

    let a_nevra = a.get_nevra()?;
    let b_nevra = b.get_nevra()?;
    println!("Comparing:");
    println!("  A: {a_nevra}");
    println!("  B: {b_nevra}");

    diff_headers(
        "Signature header",
        a.signature.get_all_entries()?,
        b.signature.get_all_entries()?,
    );

    diff_headers(
        "Main header",
        a.header.get_all_entries()?,
        b.header.get_all_entries()?,
    );

    Ok(())
}
